rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,211 @@
+- Start Date: 09-12-2022
+- Status: Proposed
+
+# Gateway refactor
+
+## Summary
+
+This proposes a full rewrite of the current gateway service. In addition, it also proposes
+changes to the way that access control to Jupyter servers is handled. Currently, each Jupyter
+server establishes a session with the user. But going forward each user will have a session
+with the gateway which will be used to control access into each Jupyter server. 
+The main motivation for changing the access control for Jupyter servers is the desire to run
+each Jupyter server in its own subdomain. The problem with unique subdomains is that each Jupyter
+server's subdomain would have to be registered in Keycloak so that the Oauth2-proxy in each 
```

Comment:
What do we mean here by subdomain? Do we think we will have an fqdn of `ssh.my-session.renkulab.io` or just `my-session.renkulab.io`?

Reply:
It would probably be something like `my-session.sessions.renkulab.io`. I don't think we need to put `ssh` in the host name. SSH would just by default be on port 22 and the notebook would be on 80 or 443. So we do not need the `ssh` part at all. But we do want to have the `sessions` part preceding the name of the sessions just to separate things out properly.

Reply:
Related to the ssh issue; have we ever discussed something like the gitpod ssh gateway which would mean hosting only a single ssh entry point for all users?

https://www.gitpod.io/docs/configure/user-settings/ssh#ssh-gateway

Reply:
Ok I had to google around a bit to find the repo for that and then find the relevant code. But it is here. So that is a possibility.

https://github.com/gitpod-io/gitpod/tree/main/components/ws-proxy/pkg/sshproxy

Reply:
The ssh gateway seems to make sense. Though is the main purpose to simplify ingress and gain visibility on active ssh connections? What I'm wondering is if it's solving something we're not realizing should be a top concern, or if it's a nice-to-have.

Reply:
I think it would be a 'less moving parts' solution, which I generally prefer - I also like the visibility dimension.

If it's something that works out of the box, then I think it would save time/effort; if it's something that requires extensive customization, then I think I guess more consideration is required.

Reply:
As discussed with @olevski, I spent a bit of time looking at this last Fri - I surveyed a few projects but boiled it down to the gitpod proxy and sshr. I spent time looking at `sshr`. It more or less works and does what it indicates; however, this solution has a fundamental limitation that it is essentially gluing two ssh sessions together and there needs to be some thought around how this can work and what the implications are. (There were issues that many projects on github in this vein assume use of ssh proxying but that is obviously some friction we don't want to push to users right now).

I looked a bit at the gitpod proxy; I built it but did not really try to get it working - for me, it was quite coupled to the gitpod ecosystem and pulled down keys from a central service, iiuc. @olevski has managed to pull out the base functionalty; however.

I set up `sshr` and managed to get a variant of this working - the project is a few years old, so I had to update some libs. However, I realized that in the demo, there were private keys both on the client end and within the proxy.

I then started to think about this a little further and I guess we would end up with the same solution as gitpod - basically, having pubkeys on the proxy and some authentication keys managed by us which are used between the proxy and the session. The purpose of these is to ensure that users cannot log into each other's sessions.

That's as far as I got on this last week; happy to discuss further if we wanna pursue this.

Reply:
One other point which was not completely clear to me - can we assume that the username on the session is constant? In the gitpod case, the username is always `gitpod`. I guess we have to assume this...

Reply:
Probably the simplest solution for now would be to have no passwords/keys in the user session and use network policies to block access to ssh on user sessions from everything except the proxy. The user's pubkey gets pushed to the proxy prob via a configmap which links session name to pubkey and if you log in to the proxy successfully, you get forwarded to the remote session.

This is not an ideal solution, but prob the quickest to implement and prob sufficiently secure for our needs.

Reply:
@Panaetius - this is specifically focused on the proxying solution (if we choose to go that way).

If the user's pubkeys are inside the session, then the proxy would need the user's private key which is not really a runner. Iiuc, in gitpod, they create another set of pub/priv keys just for access from the proxy to the session but these only live inside the gitpod context (ie they are generated by gitpod and looked up by the proxy on session initialization - the user can prob see the pubkey as it lives in the session but never sees the corresponding privkey).

If we don't use the proxying solution, then I think the natural solution is as you suggest.

Reply:
@seanrmurphy where is the user's pub key? Presumably the proxy has it but how does it get it?

Reply:
nvm, I see you said above via a configmap - but this needs to be made dynamic somehow?

Reply:
> nvm, I see you said above via a configmap - but this needs to be made dynamic somehow?

I guess a configmap is most natural; of course, it could be possible to write another service which handles this. 

If we went with a configmap approach, it would need to be dynamic.

Reply:
ok, thanks for the clarification.

> One other point which was not completely clear to me - can we assume that the username on the session is constant? In the gitpod case, the username is always gitpod. I guess we have to assume this...

Right now it's usually `jovyan` and I think there's some parts of notebooks service that depend on this, but it's not really an enforced constraint. 

Also the session URLs depend on username and commit (since you can start multiple sessions for the same user), and we discussed having an additional FQDN for SSHing that can only point to one of those sessions at a time. Users could still use the FQDN of a specific session if they want, but the persistent FQDN makes their lives easier since they wouldn't have to update the connection information every time they start a new session.

I think having an ssh proxy should be separate from the gateway refactor and is something we can add in the future. For now we'd just forward whatever traffic and it's up to the sshd in the session to do auth. I have a test project with sshd set up this way [here](https://dev.renku.ch/projects/ralf.grubenmann/ssh-into-sessions/files). Though that means that the gateway doesn't do auth for port 22 traffic. But I think this would be the fastest thing to get working, and we can add a sshgateway later on as a separate feature.

Reply:
I am resolving this because the ssh proxy is out of scope for this RFC. There will be a separate RFC for that.